Allow to specify a set of device for CUDAFuture

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56518 Add tests for CUDAFuture
* #56516 Support custom Python classes in CUDAFuture
* #56517 Add CUDA support to a user-created torch.futures.Future
* **#56515 Allow to specify a set of device for CUDAFuture**

In https://github.com/pytorch/pytorch/pull/56405 we finally found a solution to support RPC remote user functions that created/used CUDA tensors on devices that were not used by their arguments, by defining a "bounding set" of devices when constructing the agent and allowing all functions to freely use any of those devices.

We had the same exact problem with the callbacks of CUDAFuture, and in this PR I'm adopting the same exact solution: I allow to specify a set of devices when constructing a CUDAFuture, and then every callback is allowed to use any of those devices. (These devices will also be propagated to child futures).

I'm also making ProcessGroupNCCL pass these devices. I can't yet do it for TensorPipeAgent, until #56405 lands.

Differential Revision: [D27861067](https://our.internmc.facebook.com/intern/diff/D27861067/)

lw committed:
Allow to specify a set of device for CUDAFuture

In https://github.com/pytorch/pytorch/pull/56405 we finally found a solution to support RPC remote user functions that created/used CUDA tensors on devices that were not used by their arguments, by defining a "bounding set" of devices when constructing the agent and allowing all functions to freely use any of those devices.

We had the same exact problem with the callbacks of CUDAFuture, and in this PR I'm adopting the same exact solution: I allow to specify a set of devices when constructing a CUDAFuture, and then every callback is allowed to use any of those devices. (These devices will also be propagated to child futures).

I'm also making ProcessGroupNCCL pass these devices. I can't yet do it for TensorPipeAgent, until #56405 lands.

Differential Revision: [D27861067](https://our.internmc.facebook.com/intern/diff/D27861067/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/cuda/CUDAFuture.h:
```diff
@@ -44,6 +47,15 @@ struct TORCH_CUDA_CPP_API CUDAFuture final : at::ivalue::Future {
+  // that the parent future didn't use. This field is set to the value provided
+  // in the constructor and will be "inherited" by all child futures.
+  // FIXME Remove the c10::optional once the TensorPipe agent can provide this.
+  c10::optional<std::vector<c10::DeviceIndex>> devices_;
```
looks like this can be const?

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1086,8 +1086,14 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
 
   {
     at::cuda::CUDAMultiStreamGuard streamGuard(ncclStreams_[key]);
+    std::vector<c10::DeviceIndex> deviceIndices;
```
nit: add `deviceIndices.reserve(device.size())`?

mrshenli commented on aten/src/ATen/cuda/CUDAFuture.cpp:
```diff
@@ -40,19 +40,61 @@ std::vector<std::reference_wrapper<const at::DataPtr>> extractDataPtrs(
+      isCudaDeviceUsed[data_ptr.device().index()] = true;
+    }
+  }
+  std::vector<c10::DeviceIndex> device_indices;
```
camel naming?

lw replied:
I'm seeing a lot of inconsistency between snake case and camel case, hence I never know what to do. Is there some convention we're following/enforcing?

mrshenli replied:
I usually try to keep the same naming convention within the same file.

lw committed:
Update on "Allow to specify a set of device for CUDAFuture"


In https://github.com/pytorch/pytorch/pull/56405 we finally found a solution to support RPC remote user functions that created/used CUDA tensors on devices that were not used by their arguments, by defining a "bounding set" of devices when constructing the agent and allowing all functions to freely use any of those devices.

We had the same exact problem with the callbacks of CUDAFuture, and in this PR I'm adopting the same exact solution: I allow to specify a set of devices when constructing a CUDAFuture, and then every callback is allowed to use any of those devices. (These devices will also be propagated to child futures).

I'm also making ProcessGroupNCCL pass these devices. I can't yet do it for TensorPipeAgent, until #56405 lands.

Differential Revision: [D27861067](https://our.internmc.facebook.com/intern/diff/D27861067/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1086,8 +1086,15 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
     work->future_ = c10::make_intrusive<at::cuda::CUDAFuture>(
-        c10::ListType::create(c10::TensorType::get()));
+        c10::ListType::create(c10::TensorType::get()),
+        std::move(deviceIndices));
```
I think there is also a call to cuda future in ` ProcessGroupNCCL::pointToPoint`, do you want to add those changes there too?

lw replied:
Good catch, thanks!

rohan-varma commented on aten/src/ATen/cuda/CUDAFuture.cpp:
```diff
@@ -70,25 +117,33 @@ void CUDAFuture::preMarkCompletedHook(
+        excessDevices.empty(),
+        "The result contained tensors residing on device(s) ",
+        formatSetOfDevices(excessDevices),
+        " which are not among the expected device(s) ",
```
nit: mention that the user can specify these devices when constructing cuda future?

lw replied:
Uhm, that makes sense, but I'm not sure how we can do it properly: this error could also be raised when a callback returns a value on a "bad" device, but the parent future could come from a variety of places (hand-constructed, returned from ProcessGroupNCCL, or from the TensorPipeAgent) each of which has its own way of setting the supported devices, and I don't think we want to list and explain all of them here. Do you have a wording that you think would be generic but still convey the message?

rohan-varma replied:
I see, its a good point that it can come from a variety of places so probably logging the devices like we do is sufficient here.

lw committed:
Update on "Allow to specify a set of device for CUDAFuture"


In https://github.com/pytorch/pytorch/pull/56405 we finally found a solution to support RPC remote user functions that created/used CUDA tensors on devices that were not used by their arguments, by defining a "bounding set" of devices when constructing the agent and allowing all functions to freely use any of those devices.

We had the same exact problem with the callbacks of CUDAFuture, and in this PR I'm adopting the same exact solution: I allow to specify a set of devices when constructing a CUDAFuture, and then every callback is allowed to use any of those devices. (These devices will also be propagated to child futures).

I'm also making ProcessGroupNCCL pass these devices. I can't yet do it for TensorPipeAgent, until #56405 lands.

Differential Revision: [D27861067](https://our.internmc.facebook.com/intern/diff/D27861067/)

[ghstack-poisoned]

lw committed:
Update on "Allow to specify a set of device for CUDAFuture"


In https://github.com/pytorch/pytorch/pull/56405 we finally found a solution to support RPC remote user functions that created/used CUDA tensors on devices that were not used by their arguments, by defining a "bounding set" of devices when constructing the agent and allowing all functions to freely use any of those devices.

We had the same exact problem with the callbacks of CUDAFuture, and in this PR I'm adopting the same exact solution: I allow to specify a set of devices when constructing a CUDAFuture, and then every callback is allowed to use any of those devices. (These devices will also be propagated to child futures).

I'm also making ProcessGroupNCCL pass these devices. I can't yet do it for TensorPipeAgent, until #56405 lands.

Differential Revision: [D27861067](https://our.internmc.facebook.com/intern/diff/D27861067/)

[ghstack-poisoned]
